ostd/src/mm/page_table/node/mod.rs

```diff
@@ -31,74 +31,67 @@ mod entry;
-        page_table::{load_pte, store_pte},
-        FrameAllocOptions, Infallible, Paddr, PagingConstsTrait, PagingLevel, VmReader,
-    },
+use crate::mm::{
+    frame::{meta::AnyFrameMeta, Frame, FrameRef},
+    paddr_to_vaddr,
+    page_table::{load_pte, store_pte},
+    FrameAllocOptions, Infallible, PagingConstsTrait, PagingLevel, VmReader,
 };
 
-/// The raw handle to a page table node.
+/// A smart pointer to a page table node.
 ///
-/// This handle is a referencer of a page table node. Thus creating and dropping it will affect
-/// the reference count of the page table node. If dropped the raw handle as the last reference,
-/// the page table node and subsequent children will be freed.
+/// This smart pointer is an owner of a page table node. Thus creating and
+/// dropping it will affect the reference count of the page table node. If
+/// dropped it as the last reference, the page table node and subsequent
+/// children will be freed.
 ///
-/// Only the CPU or a PTE can access a page table node using a raw handle. To access the page
-/// table node from the kernel code, use the handle [`PageTableNode`].
-#[derive(Debug)]
-pub(super) struct RawPageTableNode<E: PageTableEntryTrait, C: PagingConstsTrait> {
-    raw: Paddr,
-    level: PagingLevel,
-    _phantom: PhantomData<(E, C)>,
-}
+/// [`PageTableNode`] is read-only. To modify the page table node, lock and use
+/// [`PageTableGuard`].
+pub(super) type PageTableNode<E, C> = Frame<PageTablePageMeta<E, C>>;
 
-impl<E: PageTableEntryTrait, C: PagingConstsTrait> RawPageTableNode<E, C> {
-    pub(super) fn paddr(&self) -> Paddr {
-        self.raw
+impl<E: PageTableEntryTrait, C: PagingConstsTrait> PageTableNode<E, C> {
+    pub(super) fn level(&self) -> PagingLevel {
+        self.meta().level
     }
 
-    pub(super) fn level(&self) -> PagingLevel {
-        self.level
+    pub(super) fn is_tracked(&self) -> MapTrackingStatus {
+        self.meta().is_tracked
     }
 
-    /// Converts a raw handle to an accessible handle by pertaining the lock.
-    pub(super) fn lock(self) -> PageTableNode<E, C> {
-        let level = self.level;
-        let page: Frame<PageTablePageMeta<E, C>> = self.into();
+    /// Allocates a new empty page table node.
+    ///
+    /// This function returns a locked owning guard.
+    pub(super) fn alloc(level: PagingLevel, is_tracked: MapTrackingStatus) -> Self {
+        let meta = PageTablePageMeta::new(level, is_tracked);
+        let frame = FrameAllocOptions::new()
+            .zeroed(true)
+            .alloc_frame_with(meta)
+            .expect("Failed to allocate a page table node");
+        // The allocated frame is zeroed. Make sure zero is absent PTE.
+        debug_assert!(E::new_absent().as_bytes().iter().all(|&b| b == 0));
 
-        // Acquire the lock.
-        let meta = page.meta();
-        while meta
+        frame
+    }
+
+    /// Locks the page table node.
+    pub(super) fn lock(&self) -> PageTableGuard<'_, E, C> {
```

Comment:
Safely constructing a `PageTableGuard` must require the caller to prove that it's under the RCU grace period. Something like:
```rust
    pub(super) fn lock(&'rcu self, _guard: &'rcu dyn InAtomicMode) -> PageTableGuard<'rcu, E, C> {
```